From 6e354a6efdd579f2aa8a60a20c00bb43400febfa Mon Sep 17 00:00:00 2001
From: Andreas Robinson <andr345@gmail.com>
Date: Sat, 10 Oct 2009 16:56:28 +0200
Subject: [PATCH 15/19] test: add softdep test, modprobe: fix simple bugs in
 do_softdep

Signed-off-by: Andreas Robinson <andr345@gmail.com>
(cherry picked from commit e58e266fba80e0370d2a5cffbe07adf8f7db383f)
---
 modprobe.c                       |  6 ++--
 tests/test-modprobe/28softdep.sh | 67 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 70 insertions(+), 3 deletions(-)
 create mode 100644 tests/test-modprobe/28softdep.sh

diff --git a/modprobe.c b/modprobe.c
index bdf3668..c270b16 100644
--- a/modprobe.c
+++ b/modprobe.c
@@ -692,7 +692,7 @@ static void do_softdep(const struct module_softdep *softdep,
 
        /* Modprobe pre_modnames */
 
-       for (i = 0; i < pre_modnames->cnt; i++) {
+       for (i = 0; pre_modnames && i < pre_modnames->cnt; i++) {
                /* Reverse module order if removing. */
                j = (flags & mit_remove) ? pre_modnames->cnt-1 - i : i;
 
@@ -707,11 +707,11 @@ static void do_softdep(const struct module_softdep *softdep,
 
        /* Modprobe post_modnames */
 
-       for (i = 0; i < post_modnames->cnt; i++) {
+       for (i = 0; post_modnames && i < post_modnames->cnt; i++) {
                /* Reverse module order if removing. */
                j = (flags & mit_remove) ? post_modnames->cnt-1 - i : i;
 
-               do_modprobe(pre_modnames->str[j], NULL, "", configname,
+               do_modprobe(post_modnames->str[j], NULL, "", configname,
                        dirname, warn, flags, io_flags);
        }
 }
diff --git a/tests/test-modprobe/28softdep.sh b/tests/test-modprobe/28softdep.sh
new file mode 100644
index 0000000..d999872
--- /dev/null
+++ b/tests/test-modprobe/28softdep.sh
@@ -0,0 +1,67 @@
+#! /bin/sh
+
+BITNESS=32
+
+rm -rf tests/tmp/*
+
+MODULE_DIR=tests/tmp/lib/modules/$MODTEST_UNAME
+
+# Set up modules
+mkdir -p $MODULE_DIR
+ln tests/data/$BITNESS/normal/noexport_nodep-$BITNESS.ko $MODULE_DIR/a.ko
+ln -s a.ko $MODULE_DIR/b.ko
+ln -s a.ko $MODULE_DIR/c.ko
+ln -s a.ko $MODULE_DIR/d.ko
+ln -s a.ko $MODULE_DIR/e.ko
+ln -s a.ko $MODULE_DIR/f.ko
+
+# Set up dependencies
+cat > $MODULE_DIR/modules.dep << EOF
+/lib/modules/$MODTEST_UNAME/a.ko:
+/lib/modules/$MODTEST_UNAME/b.ko:
+/lib/modules/$MODTEST_UNAME/c.ko:
+/lib/modules/$MODTEST_UNAME/d.ko:
+/lib/modules/$MODTEST_UNAME/e.ko:
+/lib/modules/$MODTEST_UNAME/f.ko:
+EOF
+
+# Test softdeps
+
+mkdir -p tests/tmp/etc/modprobe.d
+cat > tests/tmp/etc/modprobe.d/modprobe.conf << EOF
+alias a_alias a
+alias c_alias c
+softdep c --pre a_alias b --post d e
+softdep e --post f
+softdep f --pre a
+EOF
+
+# Insert-test
+
+R1=`modprobe -v c_alias 2>&1 | \
+	grep -v INIT_MODULE\: | \
+	sed -e "s/\/lib\/modules\/$MODTEST_UNAME\///g"`
+R2=`echo $R1`	# remove newlines
+[ "$R2" = "insmod a.ko insmod b.ko insmod c.ko insmod d.ko insmod e.ko insmod a.ko insmod f.ko" ]
+
+# Remove-test
+
+R1=`modprobe -v -r c_alias 2>&1 | \
+	grep -v DELETE_MODULE\: | \
+	sed -e "s/\/lib\/modules\/$MODTEST_UNAME\///g"`
+R2=`echo $R1`	# remove newlines
+[ "$R2" = "rmmod f.ko rmmod a.ko rmmod e.ko rmmod d.ko rmmod c.ko rmmod b.ko rmmod a.ko" ]
+
+# Test loop detector
+
+mkdir -p tests/tmp/etc/modprobe.d
+cat > tests/tmp/etc/modprobe.d/modprobe.conf << EOF
+alias a_alias a
+softdep a --pre b
+softdep b --pre a
+EOF
+
+R1=`modprobe -v a_alias 2>&1 | cat` # shell won't assign to R1 without no-op cat.
+# This comparison will fail if MAX_RECURSION is odd. In that case: s/b/a/g
+[ "$R1" = "FATAL: modprobe: softdep dependency loop encountered inserting b" ]
+
-- 
1.9.0

